src/backend/gl/src/device.rs

```diff
@@ -254,6 +256,27 @@ impl Device {
         Ok(())
     }
 
+    fn set_push_const_layout(
+        &self,
+        ast: &mut spirv::Ast<glsl::Target>,
+    ) -> Result<(), d::ShaderError> {
+        let resources = ast.get_shader_resources().map_err(gen_unexpected_error)?;
+
+        match self.share.info.shading_language.tuple() {
+            // Explicit uniform locations are only valid on OpenGL 4.3+
+            (major, minor) if major >= 4 && minor >= 3 => {
+                for (index, pushconst) in resources.push_constant_buffers.iter().enumerate() {
+                    let _ = ast
```

Comment:
Afaik, push_constants need to be remaped, like everything else.

https://github.com/gfx-rs/gfx/blob/58245e48a19758c5ac05f301c9cfb14cf3a138af/src/backend/gl/src/device.rs#L336

I can't imagine how this code might possibly work. This is the only place in the code which mentions push constants.

To my knowledge, multiple things would need to happen to support push constants:
 - Use spirv reflection to change PSBs to just another UBO.
 - Remap it like everything else.
 - Copy what we get from `vkCmdPushConstants` into the UBO.
 - Maybe a couple more steps that I've missed.
 - Profit!

Reply:
Push constants don't need to be remapped, spirv-cross translates push constants to a single uniform called PushConst, and we use OpenGL reflection to get the information from it.

Reply:
All UBOs must be remapped. OGL bindings are per shader program. Vulkan bindings are per stage.

Consider this, what if two different shader stages receive two different PSBs? The two PSBs must be assigned two different bindings. These two bindings must also not conflict with the bindings of any other UBO. The solution to this is to remap them like all the other UBOs.

Reply:
I don't really understand. What we have here works, but I might only be using one stage? What stages are you referring to?

Reply:
Sort've rushing to type this, forgive errors.

Multiple push constant ranges can be used in a single pipeline:

> pushConstantRangeCount is the number of push constant ranges included in the pipeline layout.
>
> pPushConstantRanges is a pointer to an array of VkPushConstantRange structures defining a set of push constant ranges for use in a single pipeline layout. In addition to descriptor set layouts, a pipeline layout also describes how many push constants can be accessed by each stage of the pipeline.

-- Vulkan specs, 13.2.2

There may only be one push constant block per stage:
> There must be no more than one push constant block statically used per shader entry point.

-- Vulkan specs, 14.5.1

Now consider this, application A marks bytes 0..8 as accessible from the vertex shader, and bytes 8..16 as accessible from the fragment shader.

Let's say their vertex shader has this:
```
layout(push_constant) uniform pushConstants {
    layout(offset = 0) float test1;
    layout(offset = 4) float test2;
} u_pushConstants;
```

And the fragment this:
```
layout(push_constant) uniform pushConstants {
    layout(offset = 8) float test3;
    layout(offset = 12) float test4;
} u_pushConstants;
```

Now, afaik (haven't tested), spriv-cross would translate them to something like this:
```
struct PushConsts
{
    float test{3,1};
    float test{4,2};
};

uniform PushConsts push;
```

In OGL, UBO binding locations are per shader program, not per shader stage. In OGL, both PCB would need to be bound to a binding which:
  A) Does not conflict with the other PCB
  B) Does not conflict with any UBOs.

For A:
If our fictional application A writes the floats F to the first PCB and F' the second PCB, we could not grant them the same binding, as they could contain different information, and we can't be binding different sets of info to the same binding. *

For B:
If our fictional application A also had four UBOs, which have reserved the binding location 1, 2, 3, and 4 (post-remapping), we would have to assign our first PCB binding 5, and the second binding 6. We couldn't assign them bindings 1 and 3, because those are used.

Now, if we treat PCBs as UBOs, and share the existing infrastructure (primarily our `DescRemapData`), we could simplify the two requirements to just one: The binding must not conflict with any other UBOs. 

Indeed, we would just grab one binding of `btype == n::BindingTypes::UniformBuffers` per PCB, and call it a day.

*Actually, instead of having separate UBOs for each PCB, we could just have one big UBO, and make the floats `test{3,4}` as offsets into that one big UBO. That would avoid having to copy the same data into multiple different UBOs if the PCBs have overlapping ranges (which is allowed). Afaik, PCBs don't have to be continuous, so this might end up being necessary. Nevertheless, the binding for this big UBO must still not conflict with the binding of any of the other UBOs.

Reply:
**It should be noted that we also remap UBOs. (I think I've said this already.)

Consider if fictional application A had two UBOs in the vertex shader (UB1 and UB2, respectively), bound to bindings 1 and 2. It also has two UBOs in the fragment shader (UB1' and UB2', respectively), also bound to bindings 1 and 2. We would in this case remap UBOs [UB1, UB2, UB1', UB2'] to bindings [1, 2, 3, 4]. That's why I state post-remapping.